In [1]:
##Ref:https://towardsdatascience.com/downloading-datasets-into-google-drive-via-google-colab-bcb1b30b0166
## If you are using the data by mounting the google drive, use the following :
from google.colab import drive
import sys

drive.mount('/content/gdrive')
base_dir_path = "gdrive/My Drive/UpGrad/Capstone"
sys.path.insert(0, base_dir_path)

Mounted at /content/gdrive


In [2]:
# import libraties
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [3]:
# Create ratings dataframe from csv
ratings_df = pd.read_csv(f"{base_dir_path}/sample30.csv")
ratings_df.head()

id            brand  ... reviews_username user_sentiment
0  AV13O1A8GV-KLJ3akUyj  Universal Music  ...           joshua       Positive
1  AV14LG0R-jtxr-f38QfS         Lundberg  ...        dorothy w       Positive
2  AV14LG0R-jtxr-f38QfS         Lundberg  ...        dorothy w       Positive
3  AV16khLE-jtxr-f38VFn              K-Y  ...          rebecca       Negative
4  AV16khLE-jtxr-f38VFn              K-Y  ...        walker557       Negative

[5 rows x 15 columns]

In [4]:
# Check the rows x cols of the dataframe
ratings_df.shape

(30000, 15)

In [5]:
# Check all the columns in the dataframe
ratings_df.columns

Index(['id', 'brand', 'categories', 'manufacturer', 'name', 'reviews_date',
       'reviews_didPurchase', 'reviews_doRecommend', 'reviews_rating',
       'reviews_text', 'reviews_title', 'reviews_userCity',
       'reviews_userProvince', 'reviews_username', 'user_sentiment'],
      dtype='object')

### **Handle Null Values**

In [6]:
ratings_df.isnull().sum()

id                          0
brand                       0
categories                  0
manufacturer              141
name                        0
reviews_date               46
reviews_didPurchase     14068
reviews_doRecommend      2570
reviews_rating              0
reviews_text                0
reviews_title             190
reviews_userCity        28071
reviews_userProvince    29830
reviews_username           63
user_sentiment              1
dtype: int64

**Observation**
- 190 ratings do not have a title.
- 63 ratings do not have a username.
- 1 rating does not have a sentiment assigned.


In [7]:
# drop rows where username is null
ratings_df.dropna(subset=["reviews_username"], how="all", inplace=True)

In [8]:
# verify there are no null usernames and products
ratings_df.isnull().sum()

id                          0
brand                       0
categories                  0
manufacturer              141
name                        0
reviews_date               40
reviews_didPurchase     14006
reviews_doRecommend      2541
reviews_rating              0
reviews_text                0
reviews_title             189
reviews_userCity        28037
reviews_userProvince    29770
reviews_username            0
user_sentiment              1
dtype: int64

In [9]:
ratings_df.shape

(29937, 15)

In [10]:
# only keep relevant fields
ratings_df = ratings_df[['name', 'reviews_username', 'reviews_rating', 'reviews_date']]
print(ratings_df.columns)
print(ratings_df.shape)

Index(['name', 'reviews_username', 'reviews_rating', 'reviews_date'], dtype='object')
(29937, 4)


In [11]:
ratings_df.head(5)

name  ...              reviews_date
0   Pink Friday: Roman Reloaded Re-Up (w/dvd)  ...  2012-11-30T06:21:45.000Z
1  Lundberg Organic Cinnamon Toast Rice Cakes  ...  2017-07-09T00:00:00.000Z
2  Lundberg Organic Cinnamon Toast Rice Cakes  ...  2017-07-09T00:00:00.000Z
3            K-Y Love Sensuality Pleasure Gel  ...  2016-01-06T00:00:00.000Z
4            K-Y Love Sensuality Pleasure Gel  ...  2016-12-21T00:00:00.000Z

[5 rows x 4 columns]

### **Drop duplicates (name, reviews_username) pairs and keep the last one**

In [12]:
# List duplicate entries
ratings_df[ratings_df.duplicated(subset=['name','reviews_username'])].head(5)

name  ...              reviews_date
2          Lundberg Organic Cinnamon Toast Rice Cakes  ...  2017-07-09T00:00:00.000Z
12                   K-Y Love Sensuality Pleasure Gel  ...  2016-04-08T00:00:00.000Z
44          Johnson's Baby Bubble Bath and Wash, 15oz  ...  2017-05-09T00:00:00.000Z
79  Olay Regenerist Deep Hydration Regenerating Cream  ...  2016-04-30T00:00:00.000Z
90  Olay Regenerist Deep Hydration Regenerating Cream  ...  2016-04-30T03:08:38.000Z

[5 rows x 4 columns]

In [13]:
# Investigate one of the duplicates
ratings_df[
  (ratings_df.name == 'Olay Regenerist Deep Hydration Regenerating Cream') & 
  (ratings_df.reviews_username == 'mylifeinheels')
]

name  ...              reviews_date
67  Olay Regenerist Deep Hydration Regenerating Cream  ...  2016-04-30T03:08:38.000Z
79  Olay Regenerist Deep Hydration Regenerating Cream  ...  2016-04-30T00:00:00.000Z
90  Olay Regenerist Deep Hydration Regenerating Cream  ...  2016-04-30T03:08:38.000Z

[3 rows x 4 columns]

In [14]:
ratings_df.shape

(29937, 4)

In [15]:
# Sort ratings_df by reviews_date column in descending order.
ratings_df = ratings_df.sort_values(by="reviews_date", ascending=False, inplace=False)
ratings_df.head(5)

name  ...              reviews_date
3471    Sunflower Swag With Metal Frame - Nearly Natural  ...  2018-01-10T19:56:37.000Z
26763  Ben & Jerry's Coffee, Coffee Buzzbuzzbuzz! Ice...  ...  2018-01-06T23:44:55.000Z
1228   Finish Quantum Dishwasher Detergent, Lemon Spa...  ...  2018-01-06T14:17:53.000Z
16773  Romantic Drama: Greatest Classic Films Collect...  ...  2017-12-26T21:06:40.000Z
18549                                   Equals (blu-Ray)  ...  2017-12-26T20:18:47.000Z

[5 rows x 4 columns]

In [16]:
# drop duplicates and save in after_drop_df
after_drop_df = ratings_df.drop_duplicates(subset=['name','reviews_username'], keep="first")
after_drop_df.head()

name  ...              reviews_date
3471    Sunflower Swag With Metal Frame - Nearly Natural  ...  2018-01-10T19:56:37.000Z
26763  Ben & Jerry's Coffee, Coffee Buzzbuzzbuzz! Ice...  ...  2018-01-06T23:44:55.000Z
1228   Finish Quantum Dishwasher Detergent, Lemon Spa...  ...  2018-01-06T14:17:53.000Z
16773  Romantic Drama: Greatest Classic Films Collect...  ...  2017-12-26T21:06:40.000Z
18549                                   Equals (blu-Ray)  ...  2017-12-26T20:18:47.000Z

[5 rows x 4 columns]

In [17]:
print(ratings_df.shape)
print(after_drop_df.shape)

(29937, 4)
(27588, 4)


In [18]:
# verify duplicate is dropped
after_drop_df[
  (ratings_df.name == 'Olay Regenerist Deep Hydration Regenerating Cream') & 
  (ratings_df.reviews_username == 'mylifeinheels')
]

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:4: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  after removing the cwd from sys.path.


name  ...              reviews_date
90  Olay Regenerist Deep Hydration Regenerating Cream  ...  2016-04-30T03:08:38.000Z

[1 rows x 4 columns]

In [19]:
ratings_df.drop_duplicates(subset=['name','reviews_username'], keep="first", inplace=True)
ratings_df.shape

(27588, 4)

In [20]:
# Drop date col
ratings_df.drop(columns='reviews_date', inplace=True)
ratings_df.head(5)

name  ... reviews_rating
3471    Sunflower Swag With Metal Frame - Nearly Natural  ...              2
26763  Ben & Jerry's Coffee, Coffee Buzzbuzzbuzz! Ice...  ...              5
1228   Finish Quantum Dishwasher Detergent, Lemon Spa...  ...              4
16773  Romantic Drama: Greatest Classic Films Collect...  ...              5
18549                                   Equals (blu-Ray)  ...              5

[5 rows x 3 columns]

In [21]:
ratings_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 27588 entries, 3471 to 28713
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   name              27588 non-null  object
 1   reviews_username  27588 non-null  object
 2   reviews_rating    27588 non-null  int64 
dtypes: int64(1), object(2)
memory usage: 862.1+ KB


In [22]:
# rename columns 
ratings_df.rename(columns={
    "name": "product",
    "reviews_username": "username",
    "reviews_rating": "rating"
}, inplace=True)

ratings_df.head()

product       username  rating
3471    Sunflower Swag With Metal Frame - Nearly Natural  bostonshopper       2
26763  Ben & Jerry's Coffee, Coffee Buzzbuzzbuzz! Ice...  coffee dreams       5
1228   Finish Quantum Dishwasher Detergent, Lemon Spa...       emmy1977       4
16773  Romantic Drama: Greatest Classic Films Collect...            msd       5
18549                                   Equals (blu-Ray)             cn       5

In [23]:
# Let's check the unique number of values in each column. Mostly checking for unique products
ratings_df.nunique()

product       271
username    24914
rating          5
dtype: int64

## Dividing the dataset into train and test

In [24]:
# Test and Train split of the dataset.
from sklearn.model_selection import train_test_split
train, test = train_test_split(ratings_df, test_size=0.30, random_state=31)

In [25]:
print(train.shape)
print(test.shape)

(19311, 3)
(8277, 3)


In [26]:
# Pivot the train ratings' dataset into matrix format in which columns are products and the rows are usernames.
df_pivot = train.pivot(
    index='username',
    columns='product',
    values='rating'
).fillna(0)

df_pivot.head(3)

product   0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. Fire File Chest  ...  Yes To Grapefruit Rejuvenating Body Wash
username                                                                 ...                                          
00dog3                                                  0.0              ...                                       0.0
00sab00                                                 0.0              ...                                       0.0
01impala                                                0.0              ...                                       0.0

[3 rows x 258 columns]

In [27]:
train.head(20)

product      username  rating
14671  Clorox Disinfecting Wipes Value Pack Scented 1...      1stgrade       5
27190                     Vicks Vaporub, Regular, 3.53oz  1friendlycat       5
2169        Mike Dave Need Wedding Dates (dvd + Digital)     twoface27       5
18201                 There's Something About Mary (dvd)      kmac1982       5
27898  Planes: Fire Rescue (2 Discs) (includes Digita...       jojo831       5
2693   My Big Fat Greek Wedding 2 (blu-Ray + Dvd + Di...          mija       4
24350  Godzilla 3d Includes Digital Copy Ultraviolet ...      tkossako       4
15850                    Burt's Bees Lip Shimmer, Raisin      ekmacula       4
24067  Godzilla 3d Includes Digital Copy Ultraviolet ...     bigbullbc       4
22655  Clear Scalp & Hair Therapy Total Care Nourishi...        stacey       5
15940                    Burt's Bees Lip Shimmer, Raisin       kmeezer       5
231    Olay Regenerist Deep Hydration Regenerating Cream         lorac       5
23907  Godzilla 3d Includes Digital Copy Ultraviolet ...   screamforme       4
10509  Clorox Disinfecting Wipes Value Pack Scented 1...          kp41       5
11655  Clorox Disinfecting Wipes Value Pack Scented 1...      tawnarue       5
28325  Planes: Fire Rescue (2 Discs) (includes Digita...       panch80       5
10879  Clorox Disinfecting Wipes Value Pack Scented 1...      lawalk01       5
8678   Clorox Disinfecting Wipes Value Pack Scented 1...          soph       5
4361   Hoover174 Platinum Collection153 Lightweight B...        nicole       5
20229               Clorox Disinfecting Bathroom Cleaner          tony       5

In [28]:
df_pivot.loc['stacey'].sum()

11.0

### Creating dummy train & dummy test dataset
These dataset will be used for prediction 
- Dummy train will be used later for prediction of the movies which has not been rated by the user. To ignore the movies rated by the user, we will mark it as 0 during prediction. The movies not rated by user is marked as 1 for prediction in dummy train dataset. 

- Dummy test will be used for evaluation. To evaluate, we will only make prediction on the movies rated by the user. So, this is marked as 1. This is just opposite of dummy_train.

In [29]:
# Copy the train dataset into dummy_train
dummy_train = train.copy()

In [30]:
# The movies not rated by user is marked as 1 for prediction. 
dummy_train['rating'] = dummy_train['rating'].apply(lambda x: 0 if x>=1 else 1)

In [31]:
# Convert the dummy train dataset into matrix format.
# set 1 for all unrated products for user
dummy_train = dummy_train.pivot(
    index='username',
    columns='product',
    values='rating'
).fillna(1)

In [32]:
dummy_train.head()

product   0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. Fire File Chest  ...  Yes To Grapefruit Rejuvenating Body Wash
username                                                                 ...                                          
00dog3                                                  1.0              ...                                       1.0
00sab00                                                 1.0              ...                                       1.0
01impala                                                1.0              ...                                       1.0
02dakota                                                1.0              ...                                       1.0
02deuce                                                 1.0              ...                                       1.0

[5 rows x 258 columns]

In [33]:
dummy_train.shape[1] - dummy_train.loc['00sab00'].sum()

2.0

**Cosine Similarity**

Cosine Similarity is a measurement that quantifies the similarity between two vectors [Which is Rating Vector in this case] 

**Adjusted Cosine**

Adjusted cosine similarity is a modified version of vector-based similarity where we incorporate the fact that different users have different ratings schemes. In other words, some users might rate items highly in general, and others might give items lower ratings as a preference. To handle this nature from rating given by user , we subtract average ratings for each user from each user's rating for different movies.



# User Similarity Matrix

## Using Cosine Similarity

In [34]:
from sklearn.metrics.pairwise import pairwise_distances

# Creating the User Similarity Matrix using pairwise_distance function.
user_correlation = 1 - pairwise_distances(df_pivot, metric='cosine')
user_correlation[np.isnan(user_correlation)] = 0
print(user_correlation)

[[1.        0.        0.        ... 0.        0.        0.       ]
 [0.        1.        0.        ... 0.        0.9486833 0.       ]
 [0.        0.        1.        ... 0.        0.        0.       ]
 ...
 [0.        0.        0.        ... 1.        0.        1.       ]
 [0.        0.9486833 0.        ... 0.        1.        0.       ]
 [0.        0.        0.        ... 1.        0.        1.       ]]


In [35]:
user_correlation.shape

(17895, 17895)

## Using adjusted Cosine 

### Here, we are not removing the NaN values and calculating the mean only for the movies rated by the user

In [36]:
# Create a user-movie matrix.
df_pivot = train.pivot(
    index='username',
    columns='product',
    values='rating'
)
df_pivot.head()

product   0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. Fire File Chest  ...  Yes To Grapefruit Rejuvenating Body Wash
username                                                                 ...                                          
00dog3                                                  NaN              ...                                       NaN
00sab00                                                 NaN              ...                                       NaN
01impala                                                NaN              ...                                       NaN
02dakota                                                NaN              ...                                       NaN
02deuce                                                 NaN              ...                                       NaN

[5 rows x 258 columns]

### Normalising the rating of the movie for each user around 0 mean

In [37]:
mean = np.nanmean(df_pivot, axis=1)
df_subtracted = (df_pivot.T-mean).T

In [38]:
df_subtracted.head()

product   0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. Fire File Chest  ...  Yes To Grapefruit Rejuvenating Body Wash
username                                                                 ...                                          
00dog3                                                  NaN              ...                                       NaN
00sab00                                                 NaN              ...                                       NaN
01impala                                                NaN              ...                                       NaN
02dakota                                                NaN              ...                                       NaN
02deuce                                                 NaN              ...                                       NaN

[5 rows x 258 columns]

In [39]:
df_pivot.sum().sort_values(ascending=False)

product
Clorox Disinfecting Wipes Value Pack Scented 150 Ct Total              26034.0
Godzilla 3d Includes Digital Copy Ultraviolet 3d/2d Blu-Ray/dvd         9512.0
Clorox Disinfecting Bathroom Cleaner                                    6499.0
Planes: Fire Rescue (2 Discs) (includes Digital Copy) (blu-Ray/dvd)     3717.0
Burt's Bees Lip Shimmer, Raisin                                         2857.0
                                                                        ...   
Lite Source Reiko 1 Light Table Lamp - Orange                              3.0
Wallmount Server Cabinet (450mm, 9 RU)                                     3.0
Sunflower Swag With Metal Frame - Nearly Natural                           2.0
Pocket Watch Wall Clock Distressed Black - Yosemite Home Decor174          1.0
Udi's Pepperoni Pizza                                                      1.0
Length: 258, dtype: float64

In [40]:
col_name = "Clorox Disinfecting Wipes Value Pack Scented 150 Ct Total"
df_pivot[[col_name]].sort_values(by=col_name, ascending=False).head(20)

product      Clorox Disinfecting Wipes Value Pack Scented 150 Ct Total
username                                                              
0325home                                                   5.0        
minnie23                                                   5.0        
missannie                                                  5.0        
missa92                                                    5.0        
miri                                                       5.0        
miraclemom                                                 5.0        
mirabreano                                                 5.0        
mippie62                                                   5.0        
minniejo                                                   5.0        
minnie01                                                   5.0        
mimi1971                                                   5.0        
mindy208                                                   5.0        
mind                                                       5.0        
mina80                                                     5.0        
mimsay                                                     5.0        
mimmie                                                     5.0        
mimi671                                                    5.0        
mimi50                                                     5.0        
missaye1231                                                5.0        
missc                                                      5.0

In [41]:
df_subtracted.head(20)

product      0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. Fire File Chest  ...  Yes To Grapefruit Rejuvenating Body Wash
username                                                                    ...                                          
00dog3                                                     NaN              ...                                       NaN
00sab00                                                    NaN              ...                                       NaN
01impala                                                   NaN              ...                                       NaN
02dakota                                                   NaN              ...                                       NaN
02deuce                                                    NaN              ...                                       NaN
0325home                                                   NaN              ...                                       NaN
06stidriver                                                NaN              ...                                       NaN
08dallas                                                   NaN              ...                                       NaN
09mommy11                                                  NaN              ...                                       NaN
1.11E+24                                                   NaN              ...                                       NaN
1085                                                       NaN              ...                                       NaN
10ten                                                      NaN              ...                                       NaN
1143mom                                                    NaN              ...                                       NaN
11677j                                                     NaN              ...                                       NaN
1234                                                       NaN              ...                                       NaN
1234561                                                    NaN              ...                                       NaN
1234asdf                                                   NaN              ...                                       NaN
123charlie                                                 NaN              ...                                       NaN
123rs                                                      NaN              ...                                       NaN
127726                                                     NaN              ...                                       NaN

[20 rows x 258 columns]

In [42]:
df_pivot.head(20)

product      0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. Fire File Chest  ...  Yes To Grapefruit Rejuvenating Body Wash
username                                                                    ...                                          
00dog3                                                     NaN              ...                                       NaN
00sab00                                                    NaN              ...                                       NaN
01impala                                                   NaN              ...                                       NaN
02dakota                                                   NaN              ...                                       NaN
02deuce                                                    NaN              ...                                       NaN
0325home                                                   NaN              ...                                       NaN
06stidriver                                                NaN              ...                                       NaN
08dallas                                                   NaN              ...                                       NaN
09mommy11                                                  NaN              ...                                       NaN
1.11E+24                                                   NaN              ...                                       NaN
1085                                                       NaN              ...                                       NaN
10ten                                                      NaN              ...                                       NaN
1143mom                                                    NaN              ...                                       NaN
11677j                                                     NaN              ...                                       NaN
1234                                                       NaN              ...                                       NaN
1234561                                                    NaN              ...                                       NaN
1234asdf                                                   NaN              ...                                       NaN
123charlie                                                 NaN              ...                                       NaN
123rs                                                      NaN              ...                                       NaN
127726                                                     NaN              ...                                       NaN

[20 rows x 258 columns]

### Finding cosine similarity

In [43]:
from sklearn.metrics.pairwise import pairwise_distances

In [44]:
# Creating the User Similarity Matrix using pairwise_distance function.
user_correlation = 1 - pairwise_distances(df_subtracted.fillna(0), metric='cosine')
user_correlation[np.isnan(user_correlation)] = 0
print(user_correlation)

[[0. 0. 0. ... 0. 0. 0.]
 [0. 1. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]]


## Prediction - User User

Doing the prediction for the users which are positively related with other users, and not the users which are negatively related as we are interested in the users which are more similar to the current users. So, ignoring the correlation for values less than 0. 

In [45]:
user_correlation[user_correlation<0]=0
user_correlation

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 1., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

Rating predicted by the user (for movies rated as well as not rated) is the weighted sum of correlation with the movie rating (as present in the rating dataset). 

In [46]:
user_predicted_ratings = np.dot(user_correlation, df_pivot.fillna(0))
user_predicted_ratings

array([[0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 6.16541513, 0.        , ..., 2.01745773, 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       ...,
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ]])

In [47]:
user_predicted_ratings.shape

(17895, 258)

Since we are interested only in the products not rated by the user, we will ignore the products rated by the user by making it zero. 

In [48]:
dummy_train.head()

product   0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. Fire File Chest  ...  Yes To Grapefruit Rejuvenating Body Wash
username                                                                 ...                                          
00dog3                                                  1.0              ...                                       1.0
00sab00                                                 1.0              ...                                       1.0
01impala                                                1.0              ...                                       1.0
02dakota                                                1.0              ...                                       1.0
02deuce                                                 1.0              ...                                       1.0

[5 rows x 258 columns]

In [49]:
user_predicted_ratings

array([[0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 6.16541513, 0.        , ..., 2.01745773, 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       ...,
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ]])

In [50]:
user_final_rating = np.multiply(user_predicted_ratings,dummy_train)
user_final_rating.head()

product   0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. Fire File Chest  ...  Yes To Grapefruit Rejuvenating Body Wash
username                                                                 ...                                          
00dog3                                                  0.0              ...                                       0.0
00sab00                                                 0.0              ...                                       0.0
01impala                                                0.0              ...                                       0.0
02dakota                                                0.0              ...                                       0.0
02deuce                                                 0.0              ...                                       0.0

[5 rows x 258 columns]

### Finding the top 20 recommendation for the *user*

In [51]:
# Take the user ID as input.
# user_input = input("Enter your user name")
user_input = '00sab00' # TODO
print(user_input)

00sab00


In [52]:
user_final_rating.head(2)

product   0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. Fire File Chest  ...  Yes To Grapefruit Rejuvenating Body Wash
username                                                                 ...                                          
00dog3                                                  0.0              ...                                       0.0
00sab00                                                 0.0              ...                                       0.0

[2 rows x 258 columns]

In [53]:
# user_final_rating.index.tolist()

In [54]:
# user_input = '00sab00'
d = user_final_rating.loc[user_input].sort_values(ascending=False)[0:20]
d

product
Godzilla 3d Includes Digital Copy Ultraviolet 3d/2d Blu-Ray/dvd                    51.551128
Planes: Fire Rescue (2 Discs) (includes Digital Copy) (blu-Ray/dvd)                33.492445
The Resident Evil Collection 5 Discs (blu-Ray)                                     18.109382
Clorox Disinfecting Wipes Value Pack Scented 150 Ct Total                          10.901189
Jason Aldean - They Don't Know                                                     10.467618
Red (special Edition) (dvdvideo)                                                    8.380565
Clorox Disinfecting Bathroom Cleaner                                                7.187438
100:Complete First Season (blu-Ray)                                                 6.165415
Dark Shadows (includes Digital Copy) (ultraviolet) (dvdvideo)                       5.468748
Alex Cross (dvdvideo)                                                               5.357589
Tostitos Bite Size Tortilla Chips                             

# Evaluation - User User 

Evaluation will we same as you have seen above for the prediction. The only difference being, you will evaluate for the movie already rated by the user insead of predicting it for the movie not rated by the user. 

In [55]:
print(train.shape)
print(test.shape)

(19311, 3)
(8277, 3)


In [56]:
# Not common 
not_common = test[~test.username.isin(train.username)]
not_common.shape

(7164, 3)

Basically all users in test are also present in train

In [57]:
# Find out the common users of test and train dataset.
common = test[test.username.isin(train.username)]
common.shape

(1113, 3)

In [58]:
common.head()

product username  rating
21619          Nexxus Exxtra Gel Style Creation Sculptor     thai       1
20725  Kendall Comforter And Sheet Set (twin) Aqua - 7pc    julie       3
6109                    Red (special Edition) (dvdvideo)    laura       5
7615   Clorox Disinfecting Wipes Value Pack Scented 1...   justin       4
5560          Australian Gold Exotic Blend Lotion, SPF 4   rachel       5

In [59]:
# convert into the user-movie matrix.
common_user_based_matrix = common.pivot_table(index='username', columns='product', values='rating')
common_user_based_matrix.head()

product   100:Complete First Season (blu-Ray)  ...  Yes To Grapefruit Rejuvenating Body Wash
username                                       ...                                          
1234                                      NaN  ...                                       NaN
1943                                      NaN  ...                                       NaN
abbey                                     NaN  ...                                       NaN
abby                                      NaN  ...                                       NaN
acg1                                      NaN  ...                                       NaN

[5 rows x 117 columns]

In [60]:
# Convert the user_correlation matrix into dataframe.
user_correlation_df = pd.DataFrame(user_correlation)
user_correlation_df.head()

0      1      2      3      4      ...  17890  17891  17892  17893  17894
0    0.0    0.0    0.0    0.0    0.0  ...    0.0    0.0    0.0    0.0    0.0
1    0.0    1.0    0.0    0.0    0.0  ...    0.0    0.0    0.0    0.0    0.0
2    0.0    0.0    0.0    0.0    0.0  ...    0.0    0.0    0.0    0.0    0.0
3    0.0    0.0    0.0    0.0    0.0  ...    0.0    0.0    0.0    0.0    0.0
4    0.0    0.0    0.0    0.0    0.0  ...    0.0    0.0    0.0    0.0    0.0

[5 rows x 17895 columns]

In [61]:
df_subtracted.head(1)

product   0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. Fire File Chest  ...  Yes To Grapefruit Rejuvenating Body Wash
username                                                                 ...                                          
00dog3                                                  NaN              ...                                       NaN

[1 rows x 258 columns]

In [62]:
user_correlation_df['username'] = df_subtracted.index
user_correlation_df.set_index('username',inplace=True)

# col_names_map = {i:df_subtracted.index[i] for i in range(len(df_subtracted.index))}
# user_correlation_df.rename(columns=col_names_map, inplace=True)

user_correlation_df.head()

0      1      2      3      4      ...  17890  17891  17892  17893  17894
username                                     ...                                   
00dog3      0.0    0.0    0.0    0.0    0.0  ...    0.0    0.0    0.0    0.0    0.0
00sab00     0.0    1.0    0.0    0.0    0.0  ...    0.0    0.0    0.0    0.0    0.0
01impala    0.0    0.0    0.0    0.0    0.0  ...    0.0    0.0    0.0    0.0    0.0
02dakota    0.0    0.0    0.0    0.0    0.0  ...    0.0    0.0    0.0    0.0    0.0
02deuce     0.0    0.0    0.0    0.0    0.0  ...    0.0    0.0    0.0    0.0    0.0

[5 rows x 17895 columns]

In [63]:
common.head(1)

product username  rating
21619  Nexxus Exxtra Gel Style Creation Sculptor     thai       1

In [64]:
list_name = common.username.tolist()

user_correlation_df.columns = df_subtracted.index.tolist()

# train/test common users
user_correlation_df_1 =  user_correlation_df[user_correlation_df.index.isin(list_name)]

In [65]:
user_correlation_df_1.shape

(921, 17895)

In [66]:
user_correlation_df_2 = user_correlation_df_1.T[user_correlation_df_1.T.index.isin(list_name)]

In [67]:
user_correlation_df_3 = user_correlation_df_2.T

In [68]:
user_correlation_df_3.head()

1234  1943  abbey  abby  acg1  ...  xmom  yohnie1  zach  zipper  zippy
username                                 ...                                    
1234       0.0   0.0    0.0   0.0   0.0  ...   0.0      0.0   0.0     0.0    0.0
1943       0.0   0.0    0.0   0.0   0.0  ...   0.0      0.0   0.0     0.0    0.0
abbey      0.0   0.0    0.0   0.0   0.0  ...   0.0      0.0   0.0     0.0    0.0
abby       0.0   0.0    0.0   1.0   0.0  ...   0.0      0.0   0.0     0.0    0.0
acg1       0.0   0.0    0.0   0.0   0.0  ...   0.0      0.0   0.0     0.0    0.0

[5 rows x 921 columns]

In [69]:
user_correlation_df_3.shape

(921, 921)

In [70]:
user_correlation_df_3[user_correlation_df_3<0]=0

common_user_predicted_ratings = np.dot(user_correlation_df_3, common_user_based_matrix.fillna(0))
common_user_predicted_ratings

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [71]:
dummy_test = common.copy()

dummy_test['rating'] = dummy_test['rating'].apply(lambda x: 1 if x>=1 else 0)

dummy_test = dummy_test.pivot_table(index='username', columns='product', values='rating').fillna(0)

In [72]:
dummy_test.shape

(921, 117)

In [73]:
common_user_predicted_ratings = np.multiply(common_user_predicted_ratings,dummy_test)

In [74]:
common_user_predicted_ratings.head(2)

product   100:Complete First Season (blu-Ray)  ...  Yes To Grapefruit Rejuvenating Body Wash
username                                       ...                                          
1234                                      0.0  ...                                       0.0
1943                                      0.0  ...                                       0.0

[2 rows x 117 columns]

Calculating the RMSE for only the movies rated by user. For RMSE, normalising the rating to (1,5) range.

In [75]:
from sklearn.preprocessing import MinMaxScaler
from numpy import *

X  = common_user_predicted_ratings.copy() 
X = X[X>0]

scaler = MinMaxScaler(feature_range=(1, 5))
print(scaler.fit(X))
y = (scaler.transform(X))

print(y)

MinMaxScaler(copy=True, feature_range=(1, 5))
[[nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]
 ...
 [nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]]


/usr/local/lib/python3.7/dist-packages/sklearn/preprocessing/_data.py:375: RuntimeWarning: All-NaN slice encountered
  data_min = np.nanmin(X, axis=0)
/usr/local/lib/python3.7/dist-packages/sklearn/preprocessing/_data.py:376: RuntimeWarning: All-NaN slice encountered
  data_max = np.nanmax(X, axis=0)


In [76]:
common_ = common.pivot_table(index='username', columns='product', values='rating')

In [77]:
# Finding total non-NaN value
total_non_nan = np.count_nonzero(~np.isnan(y))

In [78]:
rmse = (sum(sum((common_ - y )**2))/total_non_nan)**0.5
print(rmse)

2.62390733020353


# Item Based Similarity

Taking the transpose of the rating matrix to normalize the rating around the mean for different movie ID. In the user based similarity, we had taken mean for each user instead of each movie. 

In [79]:
train.columns

Index(['product', 'username', 'rating'], dtype='object')

In [80]:
df_pivot = train.pivot(
    index='username',
    columns='product',
    values='rating'
).T

df_pivot.head()

username                                            00dog3  ...  zzz1127
product                                                     ...         
0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. F...     NaN  ...      NaN
100:Complete First Season (blu-Ray)                    NaN  ...      NaN
2017-2018 Brownline174 Duraflex 14-Month Planne...     NaN  ...      NaN
2x Ultra Era with Oxi Booster, 50fl oz                 NaN  ...      NaN
42 Dual Drop Leaf Table with 2 Madrid Chairs"          NaN  ...      NaN

[5 rows x 17895 columns]

Normalising the movie rating for each movie for using the Adujsted Cosine

In [81]:
mean = np.nanmean(df_pivot, axis=1)
df_subtracted = (df_pivot.T-mean).T

In [82]:
df_subtracted.head()

username                                            00dog3  ...  zzz1127
product                                                     ...         
0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. F...     NaN  ...      NaN
100:Complete First Season (blu-Ray)                    NaN  ...      NaN
2017-2018 Brownline174 Duraflex 14-Month Planne...     NaN  ...      NaN
2x Ultra Era with Oxi Booster, 50fl oz                 NaN  ...      NaN
42 Dual Drop Leaf Table with 2 Madrid Chairs"          NaN  ...      NaN

[5 rows x 17895 columns]

Finding the cosine similarity using pairwise distances approach

In [83]:
from sklearn.metrics.pairwise import pairwise_distances

In [84]:
# Item Similarity Matrix
item_correlation = 1 - pairwise_distances(df_subtracted.fillna(0), metric='cosine')
item_correlation[np.isnan(item_correlation)] = 0
print(item_correlation)

[[ 1.          0.          0.         ...  0.          0.
   0.        ]
 [ 0.          1.          0.         ... -0.00486513  0.
   0.        ]
 [ 0.          0.          0.         ...  0.          0.
   0.        ]
 ...
 [ 0.         -0.00486513  0.         ...  1.          0.
   0.        ]
 [ 0.          0.          0.         ...  0.          1.
   0.        ]
 [ 0.          0.          0.         ...  0.          0.
   1.        ]]


Filtering the correlation only for which the value is greater than 0. (Positively correlated)

In [85]:
item_correlation[item_correlation<0]=0
item_correlation

array([[1., 0., 0., ..., 0., 0., 0.],
       [0., 1., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 1., 0., 0.],
       [0., 0., 0., ..., 0., 1., 0.],
       [0., 0., 0., ..., 0., 0., 1.]])

# Prediction - Item Item

In [86]:
item_predicted_ratings = np.dot((df_pivot.fillna(0).T),item_correlation)
item_predicted_ratings

array([[0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.00039665, 0.        , ..., 0.00133595, 0.01380156,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.00274245,
        0.01702074],
       ...,
       [0.        , 0.        , 0.        , ..., 0.        , 0.00170264,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.00222658, 0.02300261,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.00136212,
        0.        ]])

In [87]:
item_predicted_ratings.shape

(17895, 258)

In [88]:
dummy_train.shape

(17895, 258)

### Filtering the rating only for the movies not rated by the user for recommendation

In [89]:
item_final_rating = np.multiply(item_predicted_ratings,dummy_train)
item_final_rating.head()

product   0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. Fire File Chest  ...  Yes To Grapefruit Rejuvenating Body Wash
username                                                                 ...                                          
00dog3                                                  0.0              ...                                  0.000000
00sab00                                                 0.0              ...                                  0.000000
01impala                                                0.0              ...                                  0.017021
02dakota                                                0.0              ...                                  0.028368
02deuce                                                 0.0              ...                                  0.022694

[5 rows x 258 columns]

### Finding the top 5 recommendation for the *user*



In [90]:
# Take the user ID as input
# user_input = input("Enter your user name")
user_input = '00sab00' # TODO
print(user_input)

00sab00


In [91]:
# Recommending the Top 20 products to the user.
d = item_final_rating.loc[user_input].sort_values(ascending=False)[0:20]
d

product
Various - Red Hot Blue:Tribute To Cole Porter (cd)                                               0.061189
D-Con Mice Bait Station - 3ct                                                                    0.043155
The Seaweed Bath Co. Argan Conditioner, Smoothing Citrus                                         0.033655
Musselman Apple Sauce, Cinnamon, 48oz                                                            0.029025
Clear Scalp & Hair Therapy Total Care Nourishing Shampoo                                         0.026438
Jolly Time Select Premium Yellow Pop Corn                                                        0.024817
Trend Lab Park Nursing Cover - Paisley                                                           0.021303
SC Johnson One Step No Buff Wax                                                                  0.019350
Bi-O-kleen Spray & Wipe All Purpose Cleaner                                                      0.019243
Hoover174 Platinum Collection153 Light

# Evaluation - Item Item

Evaluation will we same as you have seen above for the prediction. The only difference being, you will evaluate for the movie already rated by the user insead of predicting it for the movie not rated by the user. 

In [92]:
test.columns

Index(['product', 'username', 'rating'], dtype='object')

In [93]:
common =  test[test["product"].isin(train["product"])]
common.shape

(8263, 3)

In [94]:
common.head(4)

product     username  rating
8006   Clorox Disinfecting Wipes Value Pack Scented 1...  mamacitaof4       5
1630   Dark Shadows (includes Digital Copy) (ultravio...         tjab       4
21693          Nexxus Exxtra Gel Style Creation Sculptor         tdbl       1
17880  Sizzix Framelits Dies 1by Tim Holtz Bird Crazy...     nancy r.       5

In [95]:
common_item_based_matrix = common.pivot_table(index='username', columns='product', values='rating').T

In [96]:
common_item_based_matrix.shape

(203, 7927)

In [97]:
item_correlation_df = pd.DataFrame(item_correlation)

In [98]:
item_correlation_df.head(1)

0    1    2    3    4    5    6    ...  251  252  253  254  255  256  257
0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0

[1 rows x 258 columns]

In [99]:
item_correlation_df['product'] = df_subtracted.index
item_correlation_df.set_index('product',inplace=True)
item_correlation_df.head()

0    1    ...  256  257
product                                                       ...          
0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. F...  1.0  0.0  ...  0.0  0.0
100:Complete First Season (blu-Ray)                 0.0  1.0  ...  0.0  0.0
2017-2018 Brownline174 Duraflex 14-Month Planne...  0.0  0.0  ...  0.0  0.0
2x Ultra Era with Oxi Booster, 50fl oz              0.0  0.0  ...  0.0  0.0
42 Dual Drop Leaf Table with 2 Madrid Chairs"       0.0  0.0  ...  0.0  0.0

[5 rows x 258 columns]

In [100]:
list_name = common["product"].tolist()

In [101]:
item_correlation_df.columns = df_subtracted.index.tolist()

item_correlation_df_1 =  item_correlation_df[item_correlation_df.index.isin(list_name)]

In [102]:
item_correlation_df_2 = item_correlation_df_1.T[item_correlation_df_1.T.index.isin(list_name)]

item_correlation_df_3 = item_correlation_df_2.T

In [103]:
item_correlation_df_3.head()

0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. Fire File Chest  ...  Yes To Grapefruit Rejuvenating Body Wash
product                                                                                                            ...                                          
0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. F...                                                1.0              ...                                       0.0
100:Complete First Season (blu-Ray)                                                               0.0              ...                                       0.0
2017-2018 Brownline174 Duraflex 14-Month Planne...                                                0.0              ...                                       0.0
2x Ultra Era with Oxi Booster, 50fl oz                                                            0.0              ...                                       0.0
4C Grated Parmesan Cheese 100% Natural 8oz Shaker                                                 0.0              ...                                       0.0

[5 rows x 203 columns]

In [104]:
item_correlation_df_3[item_correlation_df_3<0]=0

common_item_predicted_ratings = np.dot(item_correlation_df_3, common_item_based_matrix.fillna(0))
common_item_predicted_ratings


array([[0.        , 0.        , 0.        , ..., 0.        , 0.02535033,
        0.        ],
       [0.        , 0.00198325, 0.        , ..., 0.        , 0.        ,
        0.00794561],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       ...,
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.00170264, ..., 0.        , 0.00462869,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.01542772,
        0.        ]])

In [105]:
common_item_predicted_ratings.shape

(203, 7927)

Dummy test will be used for evaluation. To evaluate, we will only make prediction on the movies rated by the user. So, this is marked as 1. This is just opposite of dummy_train



In [106]:
dummy_test = common.copy()

dummy_test['rating'] = dummy_test['rating'].apply(lambda x: 1 if x>=1 else 0)

dummy_test = dummy_test.pivot_table(index='username', columns='product', values='rating').T.fillna(0)

common_item_predicted_ratings = np.multiply(common_item_predicted_ratings,dummy_test)

The products not rated is marked as 0 for evaluation. And make the item- item matrix representaion.


In [107]:
common_ = common.pivot_table(index='username', columns='product', values='rating').T

In [108]:
from sklearn.preprocessing import MinMaxScaler
from numpy import *

X  = common_item_predicted_ratings.copy() 
X = X[X>0]

scaler = MinMaxScaler(feature_range=(1, 5))
print(scaler.fit(X))
y = (scaler.transform(X))

print(y)

MinMaxScaler(copy=True, feature_range=(1, 5))
[[nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]
 ...
 [nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]]


/usr/local/lib/python3.7/dist-packages/sklearn/preprocessing/_data.py:375: RuntimeWarning: All-NaN slice encountered
  data_min = np.nanmin(X, axis=0)
/usr/local/lib/python3.7/dist-packages/sklearn/preprocessing/_data.py:376: RuntimeWarning: All-NaN slice encountered
  data_max = np.nanmax(X, axis=0)


In [109]:
# Finding total non-NaN value
total_non_nan = np.count_nonzero(~np.isnan(y))

In [110]:
rmse = (sum(sum((common_ - y )**2))/total_non_nan)**0.5
print(rmse)

3.5765970294622638


## **Save Pickle File**

In [111]:
# Since User-User Similarity is giving better result ( lower root mean squared error ). We will save user_final_rating in a pickle file
import pickle

In [112]:
user_final_rating.head(10)

product      0.6 Cu. Ft. Letter A4 Size Waterproof 30 Min. Fire File Chest  ...  Yes To Grapefruit Rejuvenating Body Wash
username                                                                    ...                                          
00dog3                                                     0.0              ...                                       0.0
00sab00                                                    0.0              ...                                       0.0
01impala                                                   0.0              ...                                       0.0
02dakota                                                   0.0              ...                                       0.0
02deuce                                                    0.0              ...                                       0.0
0325home                                                   0.0              ...                                       0.0
06stidriver                                                0.0              ...                                       0.0
08dallas                                                   0.0              ...                                       0.0
09mommy11                                                  0.0              ...                                       0.0
1.11E+24                                                   0.0              ...                                       0.0

[10 rows x 258 columns]

In [113]:
# Save pickle file 
recommendation_pickle_filename = "user_final_rating.pickle"
recommendation_pickle_filepath = f"{base_dir_path}/{recommendation_pickle_filename}"
pickle.dump(user_final_rating, open(recommendation_pickle_filepath, 'wb'))

In [114]:
# load pickle file
recommendation_system = pickle.load(open(recommendation_pickle_filepath, 'rb'))
sentiment_analysis_model = pickle.load(open(f"{base_dir_path}/lr_sentiment_analysis_model.pickle", 'rb'))

## Test top 20 recommendation

In [115]:
# recommend
username = '00sab00'
top_20_recommendations = recommendation_system.loc[username].sort_values(ascending=False)[0:20]

In [116]:
og_ratings_df = pd.read_csv(f"{base_dir_path}/sample30.csv")
og_ratings_df.head()

id            brand  ... reviews_username user_sentiment
0  AV13O1A8GV-KLJ3akUyj  Universal Music  ...           joshua       Positive
1  AV14LG0R-jtxr-f38QfS         Lundberg  ...        dorothy w       Positive
2  AV14LG0R-jtxr-f38QfS         Lundberg  ...        dorothy w       Positive
3  AV16khLE-jtxr-f38VFn              K-Y  ...          rebecca       Negative
4  AV16khLE-jtxr-f38VFn              K-Y  ...        walker557       Negative

[5 rows x 15 columns]

In [117]:
top_20_recommendations

product
Godzilla 3d Includes Digital Copy Ultraviolet 3d/2d Blu-Ray/dvd                    51.551128
Planes: Fire Rescue (2 Discs) (includes Digital Copy) (blu-Ray/dvd)                33.492445
The Resident Evil Collection 5 Discs (blu-Ray)                                     18.109382
Clorox Disinfecting Wipes Value Pack Scented 150 Ct Total                          10.901189
Jason Aldean - They Don't Know                                                     10.467618
Red (special Edition) (dvdvideo)                                                    8.380565
Clorox Disinfecting Bathroom Cleaner                                                7.187438
100:Complete First Season (blu-Ray)                                                 6.165415
Dark Shadows (includes Digital Copy) (ultraviolet) (dvdvideo)                       5.468748
Alex Cross (dvdvideo)                                                               5.357589
Tostitos Bite Size Tortilla Chips                             

In [118]:
top_20_recommendations_list = top_20_recommendations.index.tolist()
top_20_recommendations_list

['Godzilla 3d Includes Digital Copy Ultraviolet 3d/2d Blu-Ray/dvd',
 'Planes: Fire Rescue (2 Discs) (includes Digital Copy) (blu-Ray/dvd)',
 'The Resident Evil Collection 5 Discs (blu-Ray)',
 'Clorox Disinfecting Wipes Value Pack Scented 150 Ct Total',
 "Jason Aldean - They Don't Know",
 'Red (special Edition) (dvdvideo)',
 'Clorox Disinfecting Bathroom Cleaner',
 '100:Complete First Season (blu-Ray)',
 'Dark Shadows (includes Digital Copy) (ultraviolet) (dvdvideo)',
 'Alex Cross (dvdvideo)',
 'Tostitos Bite Size Tortilla Chips',
 'Pendaflex174 Divide It Up File Folder, Multi Section, Letter, Assorted, 12/pack',
 'Hormel Chili, No Beans',
 'Holmes174 Personal Indoor Heater W/ Manual Controls White 1500w Hfh131-N',
 "Burt's Bees Lip Shimmer, Raisin",
 "Stargate (ws) (ultimate Edition) (director's Cut) (dvdvideo)",
 'Caress Moisturizing Body Bar Natural Silk, 4.75oz',
 'Coty Airspun Face Powder, Translucent Extra Coverage',
 "There's Something About Mary (dvd)",
 'Chips Ahoy! Original Ch

In [119]:
og_ratings_df[og_ratings_df.name == 'Godzilla 3d Includes Digital Copy Ultraviolet 3d/2d Blu-Ray/dvd'][['name', 'reviews_username', 'reviews_rating', 'reviews_title', 'reviews_text']]

name  ...                                       reviews_text
22884  Godzilla 3d Includes Digital Copy Ultraviolet ...  ...  could have left this one on the shelf but it h...
22885  Godzilla 3d Includes Digital Copy Ultraviolet ...  ...  This movie is much worse that expected. The 19...
22886  Godzilla 3d Includes Digital Copy Ultraviolet ...  ...  Just don't get the hype about Godzilla. Frankl...
22887  Godzilla 3d Includes Digital Copy Ultraviolet ...  ...  A little disappointed its not like the Godzill...
22888  Godzilla 3d Includes Digital Copy Ultraviolet ...  ...  It was on sale, so I bought it. After I got it...
...                                                  ...  ...                                                ...
26204  Godzilla 3d Includes Digital Copy Ultraviolet ...  ...  Very good movie una de las mejores pel?culas q...
26205  Godzilla 3d Includes Digital Copy Ultraviolet ...  ...  If you like monster movies, you'll like Godzil...
26206  Godzilla 3d Includes Digital Copy Ultraviolet ...  ...  I really enjoyed this movie. It isn't tacky li...
26207  Godzilla 3d Includes Digital Copy Ultraviolet ...  ...  This movie was simply perfect. It follows the ...
26208  Godzilla 3d Includes Digital Copy Ultraviolet ...  ...  The movie is awesome???great features and reco...

[3325 rows x 5 columns]

In [120]:
# # sentiment_analysis_model


# # load Vectorizer and test
# vectorizer_pickle = pickle.load(open(vectorizer_pickle_filepath, 'rb'))

# # load LR model and test
# loaded_lr_model = pickle.load(open(lr_pickle_filepath, 'rb'))

# def get_product_positive_sentiment_score(product_name):
#   # 1. Preprocess review
#   product_reviews = ratings_df[ratings_df.name == product_name][['reviews_title', 'reviews_text']]
#   # TODO Merge title and text
#   product_reviews['preprocessed_review'] = product_reviews.reviews_text.apply(document_preprocess)

#   # 2. Pass preprocessed review to vectorizer
#   tfidf_model = vectorizer_pickle.transform(product_reviews['preprocessed_review'])
#   tfidf_df = pd.DataFrame(tfidf_model.toarray(), columns = vectorizer_pickle.get_feature_names())

#   # 3. Make Prediction
#   y_proba = sentiment_analysis_model.predict_proba(tfidf_df)[:, 1]
#   y_pred = list(map(lambda x: 0 if x < 0.42 else 1, y_proba.tolist()))

#   # 4. Calculate Positive Sentiment Score
#   sentiment_score = y_pred.count(1)/len(y_pred)
#   return sentiment_score

# positive_sentiment_score_list = []
# for product_recommendation in top_20_recommendations_list:
#   # get sentiment score
#   sentiment_score = get_product_positive_sentiment_score()
#   # add to list
#   positive_sentiment_score_list.append(sentiment_score)

# top_20_recommendations_sorted_df = pd.DataFrame(data = {
#     "product": top_20_recommendations_list,
#     "sentiment_score": positive_sentiment_score_list
# })

# top_20_recommendations_sorted_df.sort_values(by="sentiment_score", ascending=False)

In [121]:
import string
import nltk
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer
nltk.download('punkt')
nltk.download('stopwords')
nltk.download('wordnet')

# load models
recommendation_system = pickle.load(open(recommendation_pickle_filepath, 'rb'))
sentiment_analysis_model = pickle.load(open(f"{base_dir_path}/lr_sentiment_analysis_model.pickle", 'rb'))

vectorizer_pickle_filename = "tfidf_vectorizer.pickle"
vectorizer_pickle_filepath = f"{base_dir_path}/{vectorizer_pickle_filename}"
vectorizer_pickle = pickle.load(open(vectorizer_pickle_filepath, 'rb'))
# Create a lemmatizer
lemmatizer = WordNetLemmatizer()

def document_preprocess(document):
  # convert to lowercase 
  document = document.lower()

  # tokenize into words
  words = word_tokenize(document)

  # replace puntuations with space
  for i in range (0, len(words)):
    word = words[i]
    words[i] = "".join(char if char not in set(string.punctuation) else " " for char in word)

  # remove stop words
  words = [word for word in words if word not in stopwords.words("english")]

  # lemmatize 
  words = [lemmatizer.lemmatize(word) for word in words]

  document = " ".join(words)

  return document

def get_product_positive_sentiment_score(product_name):
  # 1. Preprocess review
  product_reviews = og_ratings_df[og_ratings_df.name == product_name][['reviews_title', 'reviews_text']]
  # TODO Merge title and text
  product_reviews['preprocessed_review'] = product_reviews.reviews_text.apply(document_preprocess)

  # 2. Pass preprocessed review to vectorizer
  tfidf_model = vectorizer_pickle.transform(product_reviews['preprocessed_review'])
  tfidf_df = pd.DataFrame(tfidf_model.toarray(), columns = vectorizer_pickle.get_feature_names())

  # 3. Make Prediction
  y_proba = sentiment_analysis_model.predict_proba(tfidf_df)[:, 1]
  y_pred = list(map(lambda x: 0 if x < 0.42 else 1, y_proba.tolist()))

  # 4. Calculate Positive Sentiment Score
  sentiment_score = y_pred.count(1)/len(y_pred)
  return sentiment_score

def get_top_20_recommendations(username):
  top_20_recommendations = recommendation_system.loc[username].sort_values(ascending=False)[0:20]
  top_20_recommendations_list = top_20_recommendations.index.tolist()

  positive_sentiment_score_list = []
  for product_name in top_20_recommendations_list:
    # get sentiment score
    sentiment_score = get_product_positive_sentiment_score(product_name)
    # add to list
    positive_sentiment_score_list.append(sentiment_score)

  top_20_recommendations_sorted_df = pd.DataFrame(data = {
      "product": top_20_recommendations_list,
      "sentiment_score": positive_sentiment_score_list
  })

  return top_20_recommendations_sorted_df.sort_values(by="sentiment_score", ascending=False)

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Unzipping corpora/wordnet.zip.


In [122]:
def get_top_20_recommendations_2(username):
  # 1. Get top 20 recommendations
  top_20_product_recommendations = recommendation_system.loc[username].sort_values(ascending=False)[0:20]
  top_20_product_recommendations_list = top_20_product_recommendations.index.tolist()
  
  # 2. Preprocess all reviews for top 20 products
  top_products_reviews = og_ratings_df[og_ratings_df.name.isin(top_20_product_recommendations_list)][['name', 'reviews_title', 'reviews_text']]
  top_products_reviews['preprocessed_review'] = top_products_reviews.reviews_text.apply(document_preprocess)

  # 3. Pass the preprocessed reviews to verctorizer
  tfidf_model = vectorizer_pickle.transform(top_products_reviews['preprocessed_review'])
  tfidf_df = pd.DataFrame(tfidf_model.toarray(), columns = vectorizer_pickle.get_feature_names())

  # 3. 
  # y_proba = sentiment_analysis_model.predict_proba(tfidf_df)[:, 1]
  cutoff = 0.42
  top_products_reviews["pos_sentiment_prob"] = sentiment_analysis_model.predict_proba(tfidf_df)[:, 1]
  top_products_reviews["sentiment_pred"] = top_products_reviews['pos_sentiment_prob'].map(lambda x: 1 if x >= cutoff else 0)

  return top_products_reviews
  # y_pred = list(map(lambda x: 0 if x < 0.42 else 1, y_proba.tolist()))

  # 4. Calculate Positive Sentiment Score
  # sentiment_score = y_pred.count(1)/len(y_pred)
  # return sentiment_score

username = '00sab00'
top_20_df = get_top_20_recommendations_2(username)
top_20_df

name  ... sentiment_pred
1537   Dark Shadows (includes Digital Copy) (ultravio...  ...              1
1538   Dark Shadows (includes Digital Copy) (ultravio...  ...              0
1539   Dark Shadows (includes Digital Copy) (ultravio...  ...              0
1540   Dark Shadows (includes Digital Copy) (ultravio...  ...              1
1541   Dark Shadows (includes Digital Copy) (ultravio...  ...              0
...                                                  ...  ...            ...
28360  Planes: Fire Rescue (2 Discs) (includes Digita...  ...              1
28361  Planes: Fire Rescue (2 Discs) (includes Digita...  ...              1
28362  Planes: Fire Rescue (2 Discs) (includes Digita...  ...              1
28363  Planes: Fire Rescue (2 Discs) (includes Digita...  ...              1
28364  Planes: Fire Rescue (2 Discs) (includes Digita...  ...              1

[19586 rows x 6 columns]

In [123]:
# top_20_df["sentiment_pred"] = 0 if top_20_df["pos_sentiment_prob"] < 0.42 else 1
cutoff = 0.42
top_20_df["sentiment_pred"] = top_20_df['pos_sentiment_prob'].map(lambda x: 1 if x >= cutoff else 0) 
top_20_df

name  ... sentiment_pred
1537   Dark Shadows (includes Digital Copy) (ultravio...  ...              1
1538   Dark Shadows (includes Digital Copy) (ultravio...  ...              0
1539   Dark Shadows (includes Digital Copy) (ultravio...  ...              0
1540   Dark Shadows (includes Digital Copy) (ultravio...  ...              1
1541   Dark Shadows (includes Digital Copy) (ultravio...  ...              0
...                                                  ...  ...            ...
28360  Planes: Fire Rescue (2 Discs) (includes Digita...  ...              1
28361  Planes: Fire Rescue (2 Discs) (includes Digita...  ...              1
28362  Planes: Fire Rescue (2 Discs) (includes Digita...  ...              1
28363  Planes: Fire Rescue (2 Discs) (includes Digita...  ...              1
28364  Planes: Fire Rescue (2 Discs) (includes Digita...  ...              1

[19586 rows x 6 columns]

In [124]:
top_20_df.groupby('name')['sentiment_pred'].mean()

name
100:Complete First Season (blu-Ray)                                                0.820144
Alex Cross (dvdvideo)                                                              0.695035
Burt's Bees Lip Shimmer, Raisin                                                    0.827033
Caress Moisturizing Body Bar Natural Silk, 4.75oz                                  0.750000
Chips Ahoy! Original Chocolate Chip - Cookies - Family Size 18.2oz                 0.550725
Clorox Disinfecting Bathroom Cleaner                                               0.795488
Clorox Disinfecting Wipes Value Pack Scented 150 Ct Total                          0.748859
Coty Airspun Face Powder, Translucent Extra Coverage                               0.797468
Dark Shadows (includes Digital Copy) (ultraviolet) (dvdvideo)                      0.736842
Godzilla 3d Includes Digital Copy Ultraviolet 3d/2d Blu-Ray/dvd                    0.809023
Holmes174 Personal Indoor Heater W/ Manual Controls White 1500w Hfh131-N   

In [125]:
top_20_df.groupby('name')['sentiment_pred'].mean().sort_values(ascending=False)

name
Stargate (ws) (ultimate Edition) (director's Cut) (dvdvideo)                       0.892473
Red (special Edition) (dvdvideo)                                                   0.858631
Burt's Bees Lip Shimmer, Raisin                                                    0.827033
100:Complete First Season (blu-Ray)                                                0.820144
Godzilla 3d Includes Digital Copy Ultraviolet 3d/2d Blu-Ray/dvd                    0.809023
Planes: Fire Rescue (2 Discs) (includes Digital Copy) (blu-Ray/dvd)                0.800525
Coty Airspun Face Powder, Translucent Extra Coverage                               0.797468
Clorox Disinfecting Bathroom Cleaner                                               0.795488
There's Something About Mary (dvd)                                                 0.784615
Caress Moisturizing Body Bar Natural Silk, 4.75oz                                  0.750000
Clorox Disinfecting Wipes Value Pack Scented 150 Ct Total                  

In [126]:
top_20_df.groupby('name')['sentiment_pred'].mean().sort_values(ascending=False).index.tolist()

["Stargate (ws) (ultimate Edition) (director's Cut) (dvdvideo)",
 'Red (special Edition) (dvdvideo)',
 "Burt's Bees Lip Shimmer, Raisin",
 '100:Complete First Season (blu-Ray)',
 'Godzilla 3d Includes Digital Copy Ultraviolet 3d/2d Blu-Ray/dvd',
 'Planes: Fire Rescue (2 Discs) (includes Digital Copy) (blu-Ray/dvd)',
 'Coty Airspun Face Powder, Translucent Extra Coverage',
 'Clorox Disinfecting Bathroom Cleaner',
 "There's Something About Mary (dvd)",
 'Caress Moisturizing Body Bar Natural Silk, 4.75oz',
 'Clorox Disinfecting Wipes Value Pack Scented 150 Ct Total',
 'Dark Shadows (includes Digital Copy) (ultraviolet) (dvdvideo)',
 "Jason Aldean - They Don't Know",
 'Alex Cross (dvdvideo)',
 'Tostitos Bite Size Tortilla Chips',
 'The Resident Evil Collection 5 Discs (blu-Ray)',
 'Hormel Chili, No Beans',
 'Chips Ahoy! Original Chocolate Chip - Cookies - Family Size 18.2oz',
 'Holmes174 Personal Indoor Heater W/ Manual Controls White 1500w Hfh131-N',
 'Pendaflex174 Divide It Up File Folde

In [127]:
username = '00sab00'
top_20_df = get_top_20_recommendations(username)

In [128]:
top_20_df

product  sentiment_score
15  Stargate (ws) (ultimate Edition) (director's C...         0.892473
5                    Red (special Edition) (dvdvideo)         0.858631
14                    Burt's Bees Lip Shimmer, Raisin         0.827033
7                 100:Complete First Season (blu-Ray)         0.820144
0   Godzilla 3d Includes Digital Copy Ultraviolet ...         0.809023
1   Planes: Fire Rescue (2 Discs) (includes Digita...         0.800525
17  Coty Airspun Face Powder, Translucent Extra Co...         0.797468
6                Clorox Disinfecting Bathroom Cleaner         0.795488
18                 There's Something About Mary (dvd)         0.784615
16  Caress Moisturizing Body Bar Natural Silk, 4.75oz         0.750000
3   Clorox Disinfecting Wipes Value Pack Scented 1...         0.748859
8   Dark Shadows (includes Digital Copy) (ultravio...         0.736842
4                      Jason Aldean - They Don't Know         0.725490
9                               Alex Cross (dvdvideo)         0.695035
10                  Tostitos Bite Size Tortilla Chips         0.678030
2      The Resident Evil Collection 5 Discs (blu-Ray)         0.640237
12                             Hormel Chili, No Beans         0.627551
19  Chips Ahoy! Original Chocolate Chip - Cookies ...         0.550725
13  Holmes174 Personal Indoor Heater W/ Manual Con...         0.531250
11  Pendaflex174 Divide It Up File Folder, Multi S...         0.522581